### DISCUSSION TITLE
Major updates without minor or patch version

### DISCUSSION THREAD

**Discussion by larasika** [renovatebot/renovate#12375] [Request Help]
Hello !

So in our project we are using `rez` as package manager and want to use Renovate to update the dependencies to the latest available version.

However, we want it to only update dependencies to the latest major version without the minor and patch version.
For example, we want to update `maya-2018` to `maya-2022` (**not** `maya-2022.2`).

Is this possible ?

I've tried to configure the package rule to disable minor and patch update and use regex as the versioning for major update like the following : 
```
"packageRules": [
        {
            "matchPackagePatterns": [
                "*"
            ],
            "matchUpdateTypes": ["minor", "patch"],
            "enabled": false
        },
        {
            "matchPackagePatterns": [
                "*"
            ],
            "matchUpdateTypes": ["major"],
            "versioning": "regex:(?<major>\\d+)"
        }
    ]
```
But it's still trying to update the dependencies including the minor version.

Is there something that I should add to or remove from the configuration ?

Some clues would be highly appreciated ! Thank you in advance

**Comment by rarkins** [comment]
`extractVersion` might help. Otherwise a working example could help us to understand it better

**Reply by larasika** [reply]
Should I include `extractVersion` within one of the `packageRules` object then ?

So I have a dependency called `maya-2018` in the package file.
I want to update it to just update it to `maya-2022`. But renovate keeps trying to update it to `maya-2022.2.0`

**Reply by rarkins** [reply]
You can find some discussions about extractVersion in this result: https://github.com/renovatebot/renovate/search?q=extractversion&type=discussions

**Reply by larasika** [reply]
Unfortunately none of the discussions works in my case.
Renovate is still trying to update to `maya-2022.2.0` :/

I understand that updating `maya-2018` to `maya-2022.2.0` is considered as major update, but apparently it's not possible to only extract the major part ? Since the version that we have in the current package file is only major version `2018`

I think I can try to re-word my questions – so basically I want renovate to recognize which version pattern to use when updating a dependency based on the pattern of that dependency. So for example:
```
maya-2018 -> maya-2022
maya-2018.5 -> maya-2022.2
maya-2018.5.0 -> maya-2022.2.0
```

I know that Renovate has this concept for Docker https://docs.renovatebot.com/docker/#preservation-of-version-precision

But is it possible to also do this for other dependencies updates ?

Thanks!

**Reply by rarkins** [reply]
If you create a public reproduction repo then we can take a look

**Reply by larasika** [reply]
Unfortunately it's a bit complicated to create a public reproduction repo since the config is a bit specific to my company and the account is inside the company's Gitlab account.

From what I understand I think we need to make the other package managers (especially `rez`, since that what we're currently using) to work like `docker` package manager – which only updates the dependencies based on the pattern on their previous version like explained [here](https://docs.renovatebot.com/docker/#preservation-of-version-precision)

**Reply by larasika** [reply]
Here's the reproduction repo https://gitlab.com/larasika/test-renovate – though like I said, trying to reproduce it as close as possible to the one we have in the company's organization is difficult (now the runner doesn't even detect dependencies to update in the reproduction repo). Because we're using a specific `lookupNameTemplate` in the `renovate.json` file in the actual repo.

**Reply by rarkins** [reply]
Is it the type of problem which only occurs on GitLab?

**Reply by larasika** [reply]
Honestly I'm not sure, but I doubt it. I think this applies on every platform.

So for rez, I think this is the behavior we should change https://github.com/renovatebot/renovate/blob/main/lib/versioning/rez/index.spec.ts#L8-L9
`1.0` or `1.0.0` shouldn't be equal to `1`

**Reply by rarkins** [reply]
Are they higher or lower than `1`?

I'm not sure making them unequal is really what you need.

**Reply by larasika** [reply]
I think `1.0` and `1.0.0` should be lower than `1`? If we take a look at how it behaves in `docker` https://github.com/renovatebot/renovate/blob/main/lib/versioning/docker/index.spec.ts#L46. It looks like no matter what number comes after 1, if it has a minor or patch number then it's smaller. Like, the pattern matters more

**Reply by rarkins** [reply]
Is your requirement "treat 1.0 as higher than 1" or is it "When updating versions, match the precision of the existing value"? I think you're requirement is probably the latter, even though you're asking for the former.

**Reply by larasika** [reply]
Yes exactly, `When updating versions, match the precision of the existing value` --> this is what I need for `rez`.
I was just mentioning the former (`1` higher than `1.x` or `1.x.x`) because I was just looking at the test in `docker` as an example and saw that behavior that I don't see in `rez` tests

**Reply by rarkins** [reply]
In rez, is `1` treated as a version (1.0.0) or a range (take the latest 1.x)?

**Reply by larasika** [reply]
Looking at this https://github.com/renovatebot/renovate/blob/main/lib/versioning/rez/pattern.ts#L67 and the test https://github.com/renovatebot/renovate/blob/main/lib/versioning/rez/index.spec.ts#L9 `1` is treated as version `1.0.0`. 

Meanwhile range has different patterns https://github.com/renovatebot/renovate/blob/main/lib/versioning/rez/index.spec.ts#L9

**Reply by rarkins** [reply]
I'm asking about in rez outside Renovate. Does `1` imply exactly `1.0.0` or latest `1.x`? (In Docker it essentially means `1.x`)

**Reply by larasika** [reply]
In the [official documentation](https://github.com/nerdvegas/rez#examples) and [here](https://mottosso.com/bleeding-rez/#versioning) as well, it only explains how rez resolves the version of a "_requested_" dependency – e.g. if we want to install `python-2` then it will actually install `python-2.6.8` if `2.6.8` exists.

But nothing explicit about how it treats the "_current_" version to update.

**Reply by rarkins** [reply]
Essentially that means we should treat 1 as a range, not a version

**Reply by larasika** [reply]
So that means we should also keep the behaviour in Renovate the way it is right now ? (`foo-2` will be updated to `foo-3.2` instead of `foo-3` if `3.2` exists ?)

**Reply by rarkins** [reply]
No, because we try to do "like for like" updates of ranges, meaning `foo-2` should be updated to `foo-3`.

**Reply by rarkins** [reply]
Next step is a minimal public reproduction as per https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md

**Reply by larasika** [reply]
> No, because we try to do "like for like" updates of ranges, meaning `foo-2` should be updated to `foo-3`.

Ok alright. Yes, this is what we want. I thought you were suggesting to keep the behavior as it is since the default behavior of `rez` itself is different than what we want.

Let me try again to re-create the minimal reproduction repo without dependencies of the company's projects then.

**Reply by larasika** [reply]
Hello @rarkins ! 
I finally managed to recreate the minimal public reproduction repo. 
You can check it out here https://gitlab.com/larasika/test-renovate/-/tree/main. Let me know if anything is missing. Thank you !

**Reply by rarkins** [reply]
Thanks, can you do the following:
- Reduce it to one dependency only (assuming you can still reproduce)
- Put on a public github.com repo so that we can fork it into our reproductions org
- Create a feature request to support ranges in `rez` versioning

**Reply by larasika** [reply]
Yes, I will try to migrate it from Gitlab to Github.
Could you tell me how I can create a feature request ? Thanks !

**Reply by rarkins** [reply]
Create a new issue and select feature when you are presented with options

**Reply by larasika** [reply]
Hi @rarkins !
Sorry it took a while for me to migrate the reproduction repo from Gitlab to Github – as I'm not used to set up Renovate on Github yet.

But anyway, here goes the [reproduction repo](https://github.com/larasika/test-renovate) in Github and the [feature request](https://github.com/renovatebot/renovate/issues/12476).

When migrating the repo to Github, I tried to follow this [documentation](https://github.com/renovatebot/github-action) to set up Renovate. It might be working now because of installing Github App.
Please do let me know if there's anything missing in my setup. Thank you !
